File: python/crypto-square/crypto_square.py
from math import ceil

def encode(plain_text):
    if len(plain_text) == 0:
        return ''
    normalized = normalize(plain_text)
    if len(normalized) <= 1:
        return normalized
    col, row = break_line(len(normalized))
    arr = [normalized[c::row] for c in range(0, col)]
    encoded_str = ''.join(arr)
    if (col * row) > len(encoded_str):
        encoded_str += ' ' * (col * row - len(encoded_str))
    return ' '.join([encoded_str[r:(r + col)] for r in range(0, len(encoded_str), col)])

def normalize(plain_text):
    return ''.join([c for c in plain_text.lower() if 'a' <= c <= 'z' or '0' <= c <= '9'])

def break_line(l):
    for col in range(1, l):
        row = ceil(l / col)
        if 0 <= (row - col) <= 1:
            return (col, row)
    return None

File: haskell/crypto-square/src/CryptoSquare.hs
module CryptoSquare (encode) where

import Data.List (intercalate)
import Data.Char (isAlphaNum, toLower)
import Data.List.Split (chunksOf)

encode :: String -> String
encode "" = ""
encode xs =
  let xss = normalize xs
      (row, col) = (bound . length) xss
   in intercalate " " (reorder (breakLine row col xss))

normalize :: String -> String
normalize text = [toLower c | c <- text, isAlphaNum c ]

bound :: Int -> (Int, Int)
bound len = (head . (filter valid)) [(divCeil len col, col) | col <- [1..len]]
  where valid (r, c) = c - r >= 0 && c - r <= 1

breakLine :: Int -> Int -> String -> [String]
breakLine row col xs = chunksOf col xs

reorder :: [String] -> [String]
reorder xs
  | all null xs = []
reorder xs = [if null x then ' ' else head x | x <- xs] : reorder [if null x then [] else tail x | x <- xs]

divCeil :: Int -> Int -> Int
divCeil a b =
  let m = a `mod` b
   in if m > 0 then a `div` b + 1 else a `div` b
